sycl/doc/Array_Kernel_Parameters.md

```diff
@@ -0,0 +1,435 @@
+The callee is a clone of the SYCL kernel object.
+The caller is generated in the form of an OpenCL kernel function.
+It receives the lambda capture object in pieces, assembles the pieces
+into the original lambda capture object and then calls the callee:
+
+```C++
+spir_kernel void caller(
+    __global int* AccData, // arg1 of Accessor init function
+    range<1> AccR1,        // arg2 of Accessor init function
+    range<1> AccR2,        // arg3 of Accessor init function
+    id<1> I,               // arg4 of Accessor init function
+    int i,
+    struct S s
+)
+{
+    // Local capture object
+    struct Capture local;
+
+    // Reassemble capture object from parts
+    local.i = i;
+    local.s = s;
+    // Call accessor�s init function
+    Accessor::init(&local.outAcc, AccData, AccR1, AccR2, I);
+
+    // Call the kernel body
+    callee(&local, id<1> wi);
+}
+
+spir_func void callee(struct Capture* this, id<1> wi)
+{
+}
+```
+
+As may be observed from the example above, standard-layout lambda capture
+components are passed by value to the device as separate parameters.
+This includes scalars, pointers, and standard-layout structs.
+Certain SYCL struct types that are not standard-layout,
+such as Accessors and Samplers, are treated specially.
+The arguments to their init functions are passed as separate parameters
+and used within the kernel caller function to initialize Accessors/Samplers
+on the device by calling their init functions using the received arguments.
+
+There is one other aspect of code generation. An �integration header�
+is generated for use during host compilation.
+This header file contains entries for each kernel.
+Among the items it defines is a table of sizes and offsets of the
+kernel parameters.
+For the source example above the integration header contains the
+following snippet:
+
+```C++
+// array representing signatures of all kernels defined in the
+// corresponding source
+static constexpr
+const kernel_param_desc_t kernel_signatures[] = {
+  //--- _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE19->18clES2_E6Worker
+  { kernel_param_kind_t::kind_accessor, 4062, 0 },
+  { kernel_param_kind_t::kind_std_layout, 4, 32 },
+  { kernel_param_kind_t::kind_std_layout, 4, 36 },
+};
+```
+
+Each entry in the kernel_signatures table contains three values:
+1) an encoding of the type of capture object member
+2) a field that encodes additional properties, and
+3) an offset within a block of memory where the value of that
+4) kernel argument is placed.
+
+The previous sections described how kernel arguments are handled today.
+The next three sections describe support for arrays.
+
+<h3>Fix 1: Kernel Arguments that are Standard-Layout Arrays</h3>
+
+As described earlier, each variable captured by a lambda that comprises a
+SYCL kernel becomes a parameter of the kernel caller function.
+For arrays, simply allowing them through would result in a
+function parameter of array type. This is not supported in C++.
```

Comment:
Earlier it was mentioned that in std C++ arrays are captured. Is SYCL solution repeating the std C++ solution or not? Would be good to mention here to have a baseline for comparison and better understanding, since std C++ was mentioned.

Reply:
In C++, a special case is made for an array captured in a lambda. The whole array is captured.
For a SYCL lambda the behavior is now the same.